marlowe-playground-client/src/Marlowe/Linter.purs

```diff
@@ -138,198 +147,427 @@ termToRange a { row, column } =
+    pure (ValueSimp pos false t)
+
+lintValue env t@(Term (Scale (Term r@(Rational a b) pos2) c) pos) = do
+  sc <- lintValue env c
+  if (b == zero) then do
+    _ <- CMS.modify (over _warnings (Set.insert (DivisionByZero (termToRange r pos2))))
+    markSimplification constToVal SimplifiableValue c sc
+    pure (ValueSimp pos false t)
+  else
+    let
+      gcdv = gcd a b
 
-lintAction state (Term (Choice choiceId bounds) _) = over _holes (getHoles choiceId <> getHoles bounds) state
+      na = a `div` gcdv
 
-lintAction state (Term (Notify obs) _) = lintObservation state obs
+      nb = b `div` gcdv
 
-lintAction state hole@(Hole _ _ _) = over _holes (insertHole hole) state
+      isSimp = (abs gcdv) > one
+    in
+      case sc of
+        (ConstantSimp _ _ v) -> pure (ConstantSimp pos true (evalValue (makeEnvironment zero zero) (emptyState (Slot zero)) (S.Scale (S.Rational a b) (S.Constant v))))
+        (ValueSimp _ _ v) -> do
+          _ <-
+            if isSimp then
+              markSimplification constToVal SimplifiableValue c sc
+            else
+              pure unit
+          pure (ValueSimp pos isSimp (Term (Scale (Term (Rational na nb) pos2) c) pos))
+
+lintValue env t@(Term (Scale h@(Hole _ _ _) c) pos) = do
+  sc <- lintValue env c
+  case sc of
+    (ConstantSimp _ _ v) -> if (v == zero) then pure (ConstantSimp pos true zero) else defaultCase sc
+    _ -> defaultCase sc
+  where
+  defaultCase sc = do
+    markSimplification constToVal SimplifiableValue c sc
+    pure (ValueSimp pos false t)
 
-negativeDeposit :: Term Action -> Maybe Warning
-negativeDeposit (Term (Deposit _ _ _ value) _) = NegativeDeposit <$> negativeValue value
+lintValue env t@(Term (ChoiceValue choiceId a) pos) = do
+  _ <- CMS.modify (over _holes (getHoles choiceId))
+  pure (ValueSimp pos false t)
 
-negativeDeposit _ = Nothing
+lintValue env t@(Term SlotIntervalStart pos) = pure (ValueSimp pos false t)
 
-negativeValue :: Term Value -> Maybe IRange
-negativeValue term@(Term _ pos) = do
-  v <- constantValue term
-  guard (v < zero)
-  pure (termToRange v pos)
+lintValue env t@(Term SlotIntervalEnd pos) = pure (ValueSimp pos false t)
 
-negativeValue _ = Nothing
+lintValue env t@(Term (UseValue (Term valueId pos2)) pos) = do
+  pure (ValueSimp pos false t)
 
-constantValue :: Term Value -> Maybe BigInteger
-constantValue (Term (Constant (Term v _)) _) = Just v
+lintValue env t@(Term (UseValue hole) pos) = do
+  _ <- CMS.modify (over _holes (insertHole hole))
+  pure (ValueSimp pos false t)
 
-constantValue (Term (NegValue v) _) = negate <$> constantValue v
+lintValue env hole@(Hole _ _ pos) = do
+  _ <- CMS.modify (over _holes (insertHole hole))
+  pure (ValueSimp pos false hole)
 
-constantValue (Term (AddValue a b) _) = do
-  va <- constantValue a
-  vb <- constantValue b
-  pure (va + vb)
+collectFromTuples :: forall a b. Array (a /\ b) -> Array a /\ Array b
+collectFromTuples = foldMap (\(a /\ b) -> [ a ] /\ [ b ])
```

Comment:
I find it a shame that this function doesn't exist in Prelude somewhere

Reply:
Oh! Now that you mention, we are not using it anymore